query-algebrizer/src/clauses/mod.rs

```diff
@@ -89,6 +89,7 @@ trait Contains<K, T> {
 
 trait Intersection<K> {
     fn with_intersected_keys(&self, ks: &BTreeSet<K>) -> Self;
+    fn keep_intersected_keys(&mut self, ks: &BTreeSet<K>);
```

Comment:
You have one mutating and one non-mutating method.  Is there a compelling reason to have different API regimes?